parser/openfoodfacts_taxonomy_parser/parser/taxonomy_parser.py

```diff
@@ -196,6 +199,22 @@ def _remove_separating_line(self, data: NodeData) -> NodeData:
                 # it's an entry block, there is always a separating line
                 data.preceding_lines.pop(0)
         return data
+    
+    def _get_comments_above(self, data: NodeData, line_number: int) -> list[str]:
+        """Get comments just above the current line"""
+        comments = []
+        current_line = line_number - 1
+        while data.comments_stack and data.comments_stack[-1][0] == current_line:
+            comments.append(data.comments_stack.pop()[1])
+            current_line -= 1
+        return comments[::-1]
+    
+    def _get_parent_comments(self, data: NodeData) -> list[str]:
+        """Get comments just above the current line"""
+        parent_comments = []
+        while data.preceding_lines and data.preceding_lines[-1] != "":
+            parent_comments.append(data.preceding_lines.pop())
+        return parent_comments[::-1]
```

Comment:
nit: In my opinion, the impact of impure functions in the code can be difficult to follow (even if that is the way most of the functions in this file are written)
To write a pure function for `_get_parent_comments` you could have
```python
def _update_node_data_with_parent_comments(self, data: NodeData) -> NodeData:
    # modifying the node_data and return the modified node_data
    return modified_node_data
```
(yes it may cause a slight overhead to create a new `dataclass` but we are writing an app in Python and not a game in C++, code readability is more important here) 
or
```python
def _get_parent_comments_and_preceding_lines(self, data: NodeData) -> tuple[list[str], list[str]]:
    # calculate the new parent_comments and preceding_line
    return new_parent_comments, new_preceding_lines
```
I personally prefer the first approach
Of course, these are not the only ways to write pure functions here, but the idea still stands

Reply:
@eric-nguyen-cs I agree that pure function would be better, but it's also good to stay consistent. So I would not change it now, just for this two functions (or do we already have only a few ?).

We could also simply add a TODO comment to the file to propose a rewrite with the pattern you proposed.

Reply:
I understand wanting to stay consistent although here I believe that making the 2 functions pure only makes the code more readable, without changing the readability of the rest of the code.
I just feel that writing more readable and maintainable code should be done incrementally